Add local loading layer for context subject headings

danamansana opened the pull request:
**What's this do?**
- In the `SubjectHeadingShow` component, instead of rendering null when there are no neighboring headings, render the loading page
- Factor out the `Loading Layer` component from the `BibsList` component so it can be reused

**Why are we doing this? (w/ JIRA link if applicable)**
Ticket is SCC-1933. This is needed because of the lag receiving results from the context API, which results in the Neighboring Subject Headings box loading late. This is a jumpy and confusing experience.

**How should this be tested? / Do these changes have associated tests?**
This can be tested by navigating to the show page for any moderately large subject heading. I have been using 'China'. You should see the loading layer for a couple of seconds, and then the subject headings should replace it.

**Dependencies for merging? Releasing to production?**
No dependencies.

**Has the application documentation been updated for these changes?**
This doesn't require additional documentation.

**Did someone actually run this code to verify it works?**
PR author has run this locally.

**Possible UI/UX questions**
- What should the loading text be?
- Should we give this box a fixed height? For more highly nested cases, the Neighboring headings box can push the Related headings box down a bit when it loads (see United States -- 1783-1815 -- History -- Registers)

siberry commented on src/app/components/SubjectHeading/LocalLoadingLayer.jsx:
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+
+export default props => (
+  <div className="nypl-column-half bibsList subjectHeadingShowLoadingWrapper">
```
I know this works. But I think this is the only component using this shorthand and not explicitly naming the component. So, I'd rather follow the same pattern of the other components because this could be confusing to someone later on.

danamansana replied:
Sounds good, changed.

siberry commented on src/app/components/SubjectHeading/SubjectHeadingShow.jsx:
```diff
@@ -171,7 +172,7 @@ class SubjectHeadingShow extends React.Component {
               />
             </div>
-            : null
+            : <LocalLoadingLayer message="Loading More Subject Headings" />
```
What if `contextHeadings` is an empty array or something else false-y for some reason?  The local loading layer would misleadingly appear. The pattern used for `BibsList` is more accurate.

danamansana replied:
Empty array is strangely truthy. But point taken that we need to ensure `contextHeadings` is an array. I've added some extra validations/error handling which should make sure `contextHeadings` always resolves to an array after the initial API call. Let me know what you think. We can always go with the `BibsList` pattern but I'd like to keep this component stateless if possible.

siberry replied:
I do think the logic is a little funny and misses the point. What if only the `context` query times out or has some other error? We could still load the bibs and the related headings. Currently, none of the components would load.

siberry replied:
Also, this component would handle the `contextLoading` logic still. `BibsList` is a class component and managed the local loading logic, because currently it makes the bibs API call.

danamansana replied:
Sorry, I don't think I follow. Why would the bibs and related headings not load if the `context` query times out?

siberry replied:
On line 135, if the error state is true, only an error message is rendered.

danamansana replied:
OK now I understand what you meant about this being a pre-existing issue. Good spot.

danamansana commented on src/app/components/SubjectHeading/LocalLoadingLayer.jsx:
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+
+export default props => (
+  <div className="nypl-column-half bibsList subjectHeadingShowLoadingWrapper">
```
Good point, changed.

danamansana commented:
Most recent changes:
- Added new `NeighboringHeadingsBox` component. Created this as as class component since we anticipate it will eventually be stateful.
- `SubjectHeadingShow` now keeps track of the error and loading status for this component and passes them as props. Eventually this will be moved to the `NeighboringHeadingsBox` state.
- Added logic to display an error message in `NeighboringHeadingsBox` if there is an error
- Also, a visual change: I moved the loading layer inside the box. The ticket doesn't clearly state one way or the other. I think this looks a little better but let me know what you think.

siberry commented on src/app/components/SubjectHeading/SubjectHeadingShow.jsx:
```diff
@@ -47,7 +53,11 @@ class SubjectHeadingShow extends React.Component {
+          this.setState({
+            contextIsLoading: false,
+            contextError: true,
+            contextHeadings: [],
```
is this line necessary?
